controller.py

```diff
@@ -208,6 +214,106 @@ def add_items(self, collection_details):
 
         return feed_response(addition_feed)
 
+    def add_with_metadata(self, collection_details):
+        """Adds identifiers with their metadata to a Collection's catalog"""
+        client = authenticated_client_from_request(self._db)
+        if isinstance(client, ProblemDetail):
+            return client
+
+        collection, ignore = Collection.from_metadata_identifier(
+            self._db, collection_details
+        )
+
+        messages = []
+
+        feed = feedparser.parse(request.data)
+        entries = feed.get("entries", [])
+
+        if not client.data_source:
+            client.data_source = DataSource.lookup(self._db, client.key, autocreate=True)
+        data_source = client.data_source
+
+        for entry in entries:
+            urn = entry.get('id')
+            try:
+                identifier, ignore = Identifier.parse_urn(
+                    self._db, urn
+                )
+            except Exception as e:
+                identifier = None
+
+            if not identifier:
+                message = OPDSMessage(
+                    urn, INVALID_URN.status_code, INVALID_URN.detail
+                )
+            else:
+                status = HTTP_OK
+                description = "Already in catalog"
+
+                if identifier not in collection.catalog:
+                    collection.catalog_identifier(self._db, identifier)
+                    status = HTTP_CREATED
+                    description = "Successfully added"
+
+                message = OPDSMessage(urn, status, description)
+
+                links = entry.get("links")
+                images = [l for l in links if l.get("rel") == Hyperlink.IMAGE or l.get("rel") == Hyperlink.THUMBNAIL_IMAGE]
```

Comment:
A lot of this code is reminiscent of code from `metadata_layer.py`. Can you use the OPDSImporter class to turn the OPDS entry into a Metadata object? Or build a Metadata object some other way?